pkg/git/factory/gitfactory.go

```diff
@@ -2,43 +2,76 @@ package gitfactory
 
 import (
 	"context"
+	"fmt"
+	"path/filepath"
 
 	"github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/filewriter"
 	"github.com/aws/eks-anywhere/pkg/git"
+	"github.com/aws/eks-anywhere/pkg/git/gitclient"
 	"github.com/aws/eks-anywhere/pkg/git/gogithub"
 	"github.com/aws/eks-anywhere/pkg/git/providers/github"
 )
 
-type gitProviderFactory struct {
-	GitClient github.GitClient
-}
+type gitProviderFactory struct{}
 
-type Options struct {
-	GithubGitClient github.GitClient
+func New() *gitProviderFactory {
+	return &gitProviderFactory{}
 }
 
-func New(opts Options) *gitProviderFactory {
-	return &gitProviderFactory{GitClient: opts.GithubGitClient}
+func (g *gitProviderFactory) Build(ctx context.Context, cluster *v1alpha1.Cluster, fluxConfig *v1alpha1.FluxConfig, writer filewriter.FileWriter) (git.ProviderClient, git.Client, filewriter.FileWriter, error) {
+	var providerClient git.ProviderClient
+	var gitClient git.Client
+	var repoWriter filewriter.FileWriter
+	var err error
+
+	switch {
+	case fluxConfig.Spec.Github != nil:
+		providerClient, gitClient, repoWriter, err = g.buildGithubProvider(ctx, cluster.Name, fluxConfig.Spec.Github.Owner, fluxConfig.Spec.Github.Repository, fluxConfig.Spec.Github.Personal, writer)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("building github provider: %v", err)
+		}
+	default:
+		return nil, nil, nil, fmt.Errorf("no valid git provider in FluxConfigSpec. Spec: %v", fluxConfig)
+	}
+	return providerClient, gitClient, repoWriter, nil
 }
 
-// BuildProvider will configure and return the proper Github based on the given GitOps configuration.
-func (g *gitProviderFactory) BuildProvider(ctx context.Context, fluxConfig *v1alpha1.FluxConfigSpec) (git.Provider, error) {
+func (g *gitProviderFactory) buildGithubProvider(ctx context.Context, clusterName string, owner string, repo string, personal bool, writer filewriter.FileWriter) (git.ProviderClient, git.Client, filewriter.FileWriter, error) {
 	token, err := github.GetGithubAccessTokenFromEnv()
 	if err != nil {
-		return nil, err
+		return nil, nil, nil, err
 	}
-	auth := git.TokenAuth{Token: token, Username: fluxConfig.Github.Owner}
+	auth := git.TokenAuth{Token: token, Username: owner}
 	gogithubOpts := gogithub.Options{Auth: auth}
 	githubProviderClient := gogithub.New(ctx, gogithubOpts)
 	githubProviderOpts := github.Options{
-		Repository: fluxConfig.Github.Repository,
-		Owner:      fluxConfig.Github.Owner,
-		Personal:   fluxConfig.Github.Personal,
+		Repository: repo,
+		Owner:      owner,
+		Personal:   personal,
 	}
-	provider, err := github.New(g.GitClient, githubProviderClient, githubProviderOpts, auth)
+	provider, err := github.New(githubProviderClient, githubProviderOpts, auth)
 	if err != nil {
-		return nil, err
+		return nil, nil, nil, err
 	}
 
-	return provider, nil
+	localGitRepoPath := filepath.Join(clusterName, "git", repo)
+	client := gitclient.New(auth, localGitRepoPath, github.RepoUrl(owner, repo))
```

Comment:
shouldn't the gitclient agnostic among different providers? i would move it out of this `buildGithubProvider` method.

Reply:
it's just the URL structure that's defendant on the provider in this case, with a github-specific URL scheme